### ISSUE TITLE
Hydrated: type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'

### ISSUE THREAD

**Issue by orsenkucher** [felangel/bloc#1452]
**From our Discord**
I am trying to migrate all of my code to 5.0.0, but I am currently running into the following error for HydratedBloc `type '_InternalLinkedHashMap<dynamic, dynamic>' is not a subtype of type 'Map<String, dynamic>'` , this is happening in the `fromJson` method...

**To Reproduce**
Author of issue (me) thinks you just need two nested `Map<String, whatever>` to reproduce this issue.
But some sample code will be awesome to have, and I will write it a bit :wink:  

**Expected behavior**
Everything goes butter smooth

**Logs**
Stacktrace says it happened here:
```dart
@override
State get state {
  if (_state != null) return _state;
  try {
    final stateJson = storage.read(storageToken);
    if (stateJson == null) return super.state;
    return fromJson(Map<String, dynamic>.from(stateJson));
  } on dynamic catch (error, stackTrace) {
    onError(error, stackTrace);
    return super.state;
  }
}
```

**Comment by orsenkucher** [comment]
As you can see, we already encountered same issue before,  
therefore `Map<String, dynamic>.from(stateJson)` is used.

---

I've approached this issue today and came up changing hive's internal `readMap` method in `hive/lib/src/binary/binary_reader_impl.dart`  
Here is before: 
```dart
@override
Map readMap([int length]) {
  length ??= readUint32();
  var map = <dynamic, dynamic>{};
  for (var i = 0; i < length; i++) {
    map[read()] = read();
  }
  return map;
}
```

And here is after
```dart
@override
Map readMap([int length]) {
  var map = <dynamic, dynamic>{};

  length ??= readUint32();
  if (length < 1) return map;

  final keyType = readByte();
  if (keyType == FrameValueType.intT) {
    map = <int, dynamic>{};
  }
  if (keyType == FrameValueType.stringT) {
    map = <String, dynamic>{};
  }
  map[read(keyType)] = read();

  for (var i = 1; i < length; i++) {
    map[read()] = read();
  }

  return map;
}
```

**Comment by orsenkucher** [comment]
Worked for me! :smile_cat: :smile_cat:  
Now what should we actually do with it??

**Comment by orsenkucher** [comment]
So help wanted😛!
- Now we have sample to produce this bug: [ram231/samplehydratedbloc](https://github.com/ram231/samplehydratedbloc)
- My fork of sample with fixes: [orsenkucher/samplehydratedbloc](https://github.com/orsenkucher/samplehydratedbloc)
- My fork of hive with applied fix: [orsenkucher/hive](https://github.com/orsenkucher/hive)

**Comment by felangel** [comment]
@orsenkucher thanks for putting that together! Will take a look later today 👍

**Comment by felangel** [comment]
@orsenkucher I just opened a [pull request](https://github.com/ram231/samplehydratedbloc/pull/2) with a fix. I believe the issue is actually in the way the current application serializes and deserializes the model and not an issue in hive. Thoughts?

**Comment by orsenkucher** [comment]
@felangel that is what we suggest as quick fix for this problem, but it's very tedious work especially knowing that previously it worked fine

**Comment by orsenkucher** [comment]
And default json.encode was kind enough to avoid such error

**Comment by orsenkucher** [comment]
Maybe we can run this Map<>.from somewhere in library code, to relieve developers from it

**Comment by felangel** [comment]
Check out [this pull request](https://github.com/ram231/samplehydratedbloc/pull/3) showing how to use `json_serializable` to handle `toJson` and `fromJson` for us. This is my recommendation 👍

**Comment by ram231** [comment]
Thanks for the correction of my sample repo.

**Comment by felangel** [comment]
Fixed in #1460 and published in v6.0.0-dev.2 of `hydrated_bloc` 🎉
